webhook/resourcesemantics/validation/validation_admit.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	err = validate(ctx, resource, request)
+	if err != nil {
+		return webhook.MakeErrorStatus("validation failed: %v", err)
+	}
+
+	if err := ac.callback(ctx, request, gvk); err != nil {
+		return webhook.MakeErrorStatus("validation callback failed: %v", err)
+	}
+
+	return &admissionv1beta1.AdmissionResponse{Allowed: true}
+}
+
+// decodeRequest deserializes the old and new GenericCrds from the incoming request and sets up the context.
+// nil oldObj or newObj denote absence of `old` (create) or `new` (delete) objects.
+func (ac *reconciler) decodeRequestAndPrepareContext(
+	ctx context.Context,
+	req *admissionv1beta1.AdmissionRequest,
+	gvk schema.GroupVersionKind) (context.Context, resourcesemantics.GenericCRD, error) {
+
+	logger := logging.FromContext(ctx)
+	handler, ok := ac.handlers[gvk]
+	if !ok {
+		logger.Errorf("Unhandled kind: %v", gvk)
+		return ctx, nil, fmt.Errorf("unhandled kind: %v", gvk)
+	}
+
+	newBytes := req.Object.Raw
+	oldBytes := req.OldObject.Raw
+
+	// Decode json to a GenericCRD
+	var newObj resourcesemantics.GenericCRD
+	if len(newBytes) != 0 {
+		newObj = handler.DeepCopyObject().(resourcesemantics.GenericCRD)
+		newDecoder := json.NewDecoder(bytes.NewBuffer(newBytes))
+		if ac.disallowUnknownFields {
+			newDecoder.DisallowUnknownFields()
+		}
+		if err := newDecoder.Decode(&newObj); err != nil {
+			return ctx, nil, fmt.Errorf("cannot decode incoming new object: %v", err)
+		}
+	}
+
+	var oldObj resourcesemantics.GenericCRD
+	if len(oldBytes) != 0 {
+		oldObj = handler.DeepCopyObject().(resourcesemantics.GenericCRD)
+		oldDecoder := json.NewDecoder(bytes.NewBuffer(oldBytes))
+		if ac.disallowUnknownFields {
+			oldDecoder.DisallowUnknownFields()
+		}
+		if err := oldDecoder.Decode(&oldObj); err != nil {
+			return ctx, nil, fmt.Errorf("cannot decode incoming old object: %v", err)
+		}
+	}
+
+	// Set up the context for validation
+	if oldObj != nil {
+		if req.SubResource == "" {
+			ctx = apis.WithinUpdate(ctx, oldObj)
+		} else {
+			ctx = apis.WithinSubResourceUpdate(ctx, oldObj, req.SubResource)
+		}
+	} else {
+		ctx = apis.WithinCreate(ctx)
+	}
+	ctx = apis.WithUserInfo(ctx, &req.UserInfo)
+	ctx = context.WithValue(ctx, kubeclient.Key{}, ac.client)
+
+	if req.DryRun != nil && *req.DryRun {
+		ctx = apis.WithDryRun(ctx)
+	}
+
+	return ctx, newObj, nil
+}
+
+func validate(ctx context.Context, resource resourcesemantics.GenericCRD, req *admissionv1beta1.AdmissionRequest) error {
+	logger := logging.FromContext(ctx)
+
+	// Only run validation for supported create and update validaiton.
+	switch req.Operation {
+	case admissionv1beta1.Create, admissionv1beta1.Update:
+		// Supported verbs
+	default:
+		logger.Infof("Unhandled webhook validation operation, letting it through %v", req.Operation)
+		return nil
+	}
+
+	// None of the validators will accept a nil value for newObj.
+	if resource == nil {
+		return errMissingNewObject
+	}
+
+	if err := resource.Validate(ctx); err != nil {
```

Comment:
I'd move logging to the callsite (or just ignore, since it's being wrapped and passed up there as well) and just unify the paths into
`return resource.Validate(ctx)`.